### PR TITLE
Bidirectional A*

### PR DESCRIPTION
**Pull request by Guilyx**
Implement Bidirectional A* algorithm. (two A* searching for the last node added to the other's closed set).

### CODE REVIEW

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+            # Add it to the closed set
+            closed_set_A[c_id_A] = current_A
+            closed_set_B[c_id_B] = current_B
+
+            # expand_grid search grid based on motion model
+            for i, _ in enumerate(self.motion):
+                continue_A = False
+                continue_B = False
+
+                child_node_A = self.Node(current_A.x + self.motion[i][0],
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:131]: Please align the current like:
```
self.Node(currentA.x ...
          currentA.y ...
          currentA.cost ... 
```

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+                                 current_A.y + self.motion[i][1],
+                                 current_A.cost + self.motion[i][2], c_id_A)
+
+                child_node_B = self.Node(current_B.x + self.motion[i][0],
+                                 current_B.y + self.motion[i][1],
+                                 current_B.cost + self.motion[i][2], c_id_B)
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:137]: Same upper comment.

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                if not self.verify_node(child_node_B):
+                    continue_B = True
+
+                if n_id_A in closed_set_A:
+                    continue_A = True
+                
+                if n_id_B in closed_set_B:
+                    continue_B = True
+
+                if not(continue_A):
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:155]:
```suggestion
                if not continue_A:
```

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                    if n_id_A not in open_set_A:
+                        open_set_A[n_id_A] = child_node_A  # discovered a new node
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:157]: pep8 too long line.

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+                    else:
+                        if open_set_A[n_id_A].cost > child_node_A.cost:
+                            # This path is the best until now. record it
+                            open_set_A[n_id_A] = child_node_A
+            
+                if not(continue_B):
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:163]:
```suggestion
                if not continue_B:
```

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+                    if n_id_B not in open_set_B:
+                        open_set_B[n_id_B] = child_node_B  # discovered a new node
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:165]: pep8 too long line.

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+                print("Open set A is empty..")
+                break
+            
+            if len(open_set_B) == 0:
+                print("Open set B is empty..")
+                break
+
+            c_id_A = min(
+                open_set_A,
+                key=lambda o: open_set_A[o].cost + self.calc_heuristic(current_B,
```
[the remark this answers is not in the document]
Comment (Guilyx) [reply PathPlanning/BidirectionalAStar/bidirectional_a_star.py]: Is it okay ?

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py (hunk 4 above)
Comment (Guilyx) [reply PathPlanning/BidirectionalAStar/bidirectional_a_star.py]: Oops forgot to parse with autopep8, done now.

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+    # start and goal position
+    sx = 10.0  # [m]
+    sy = 10.0  # [m]
+    gx = 50.0  # [m]
+    gy = 50.0  # [m]
+    grid_size = 2.0  # [m]
+    robot_radius = 1.0  # [m]
+
+    # set obstable positions
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:302]: obstacle?

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+            closed_set_B[c_id_B] = current_B
+
+            # expand_grid search grid based on motion model
+            for i, _ in enumerate(self.motion):
+                continue_A = False
+                continue_B = False
+
+                child_node_A = self.Node(current_A.x + self.motion[i][0],
+                                         current_A.y + self.motion[i][1],
+                                         current_A.cost + self.motion[i][2], c_id_A)
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:129]: pep8 error.

Could you move the the last "c_id_A" to the next line?

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+
+                child_node_B = self.Node(current_B.x + self.motion[i][0],
+                                         current_B.y + self.motion[i][1],
+                                         current_B.cost + self.motion[i][2], c_id_B)
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:133]: Same as upper comment.

File: PathPlanning/BidirectionalAStar/bidirectional_a_star.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+                        if open_set_B[n_id_B].cost > child_node_B.cost:
+                            # This path is the best until now. record it
+                            open_set_B[n_id_B] = child_node_B
+
+        rx, ry = self.calc_final_path_bidir(
+            meetpointA, meetpointB, closed_set_A, closed_set_B)
+
+        return rx, ry
+
+    def calc_final_path_bidir(self, meetnode_A, meetnode_B, closed_set_A, closed_set_B):
```
Comment (AtsushiSakai) [inline PathPlanning/BidirectionalAStar/bidirectional_a_star.py:174]: calc_final_bidirectional_path?

**Comment by AtsushiSakai** [comment]
@Guilyx Thank you so much!! LGTM. I will merge it.

**Pull request merged**